File: hnswlib-spark/src/main/java/com/github/jelmer/knn/spark/DenseVectorDistanceFunctions.java
package com.github.jelmer.knn.spark;

import org.apache.spark.ml.linalg.Vector;

public final class DenseVectorDistanceFunctions {

    /**
     * Private constructor to prevent initialization.
     */
    private DenseVectorDistanceFunctions() {
    }

    /**
     * Calculates cosine distance on a dense vector.
     *
     * @param u Left vector.
     * @param v Right vector.
     * @return Cosine distance between u and v.
     */
    @SuppressWarnings("Duplicates")
    public static double cosineDistance(Vector u, Vector v) {
        double dot = 0.0f;
        double nru = 0.0f;
        double nrv = 0.0f;
        for (int i = 0; i < u.size(); i++) {
            dot += u.apply(i) * v.apply(i);
            nru += u.apply(i) * u.apply(i);
            nrv += v.apply(i) * v.apply(i);
        }

        double similarity = dot / (Math.sqrt(nru) * Math.sqrt(nrv));
        return 1 - similarity;
    }


    /**
     * Calculates inner product on a dense vector.
     *
     * @param u Left vector.
     * @param v Right vector.
     * @return Cosine distance between u and v.
     */
    @SuppressWarnings("Duplicates")
    public static double innerProduct(Vector u, Vector v) {
        double dot = 0;
        for (int i = 0; i < u.size(); i++) {
            dot += u.apply(i) * v.apply(i);
        }

        return 1 - dot;
    }


}

File: hnswlib-spark/src/main/scala/com/github/jelmerk/spark/linalg/functions/DenseVectorDistanceFunctions.scala
package com.github.jelmerk.spark.linalg.functions

import com.github.jelmerk.knn.scalalike._
import org.apache.spark.ml.linalg.DenseVector

object DenseVectorDistanceFunctions {

  /**
   * Calculates the cosine distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Cosine distance between u and v.
   */
  def cosineDistance(u: DenseVector, v: DenseVector): Double = doubleCosineDistance(u.values, v.values)

  /**
   * Calculates the inner product.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Cosine distance between u and v.
   */
  def innerProduct(u: DenseVector, v: DenseVector): Double = doubleInnerProduct(u.values, v.values)

  /**
   * Calculates the Bray Curtis distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Bray Curtis distance between u and v.
   */
  def brayCurtisDistance(u: DenseVector, v: DenseVector): Double = doubleBrayCurtisDistance(u.values, v.values)

  /**
   * Calculates the canberra distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Canberra distance between u and v.
   */
  def canberraDistance(u: DenseVector, v: DenseVector): Double = doubleCanberraDistance(u.values, v.values)

  /**
   * Calculates the correlation distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Correlation distance between u and v.
   */
  def correlationDistance(u: DenseVector, v: DenseVector): Double = doubleCorrelationDistance(u.values, v.values)

  /**
   * Calculates the euclidean distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Euclidean distance between u and v.
   */
  def euclideanDistance(u: DenseVector, v: DenseVector): Double = doubleEuclideanDistance(u.values, v.values)

  /**
   * Calculates the manhattan distance.
   *
   * @param u Left vector.
   * @param v Right vector.
   *
   * @return Manhattan distance between u and v.
   */
  def manhattanDistance(u: DenseVector, v: DenseVector): Double = doubleManhattanDistance(u.values, v.values)

}
